distributed/hdfs.py

```diff
@@ -104,6 +105,88 @@ def _read_csv(fn, executor=None, hdfs=None, lazy=False, lineterminator='\n',
         df = yield _futures_to_dask_dataframe(futures)
         raise gen.Return(df)
 
+def avro_header(f):
+    import json
+    import fastavro._reader as fa
+
+    f.seek(0)
+    header = fa.read_data(f, fa.HEADER_SCHEMA)
+    metadata = dict((k, fa.btou(v)) for k, v in fa.iteritems(header['meta']))
+    schema =  json.loads(metadata['avro.schema'])
+    codec = metadata.get('avro.codec', 'null')
+    
+    fa.acquaint_schema(schema, fa.READERS)
+    sync_marker = header['sync']
+    
+    read_block = fa.BLOCK_READERS.get(codec)
+    if not read_block:
+        raise ValueError('Unrecognized codec: %r' % codec)
+    return schema, sync_marker, f.tell(), codec
+
+def avro_body(data, schema, sync_marker, codec):
+    import fastavro._reader as fa
+
+    if not data.endswith(sync_marker):
+        ## Read delimited should keep end-of-block delimiter
+        data = data + sync_marker
+    stream = io.BytesIO(data)
+    block_count = 0
+    read_block = fa.BLOCK_READERS.get(codec)
+    while True:
+        block_count = fa.read_long(stream, None)
+        block_fo = read_block(stream)    
+        for i in fa.xrange(block_count):
+            yield fa.read_data(block_fo, schema, None)    
+        fa.skip_sync(stream, sync_marker)
```

Comment:
Beware, if we've actually copy-pasted code from another library then we'll need to include their license.  For snippets like this I find its often cheaper to rewrite code from a fresh mind than to copy directly.

What was off about their iterator?